src/WebJobs.Extensions.DurableTask/ETWEventListener.cs

```diff
@@ -0,0 +1,27 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    internal class ETWEventListener : EventListener
```

Comment:
Two things:
1. The name of this class doesn't look right. There shouldn't be any "ETW" involved at all in Linux scenarios.
2. I'm not familiar with the `EventListener` class. Does it work in Linux, and how is its `OnEventWritten` method (below) getting invoked?

Reply:
@cgillum, the way that I understand the way EventListener and how it works, is that it allows you to listen to "ETW" events emitted by EventSource implementations, even if actual ETW events aren't being emitted to the OS like they are on Windows. 

This allows you to set up a listener for these  "ETW" events, and we can instead write our ETW-equivalent log.

This is at least my understanding of the class, but @davidmrdavid, it may be worth spinning up a Linux VM and testing this code path there to ensure EventListener is behaving as expected in non-Windows environments.

Reply:
From what I could grasp reading around (see [here](https://medium.com/criteo-labs/c-in-process-clr-event-listeners-with-net-core-2-2-ef4075c14e87) and [here](https://github.com/aspnet/Logging/issues/572)), EventSource implementations on Linux write to LTTng, and EventListener on Linux listens to LTTng events and calls `OnEventWritten` based on those.

Because of this, I would second Chris's first critique, and we should just call this `EventSourceListener`.

Reply:
@cgillum 

I'll have to answer these two backwards for clarity.

(2) 

My understanding is that `EventListener`'s `OnEventWritten` method gets invoked every time an `EventSource` write event is emitted. So, for us, that would be a callback resulting from executing any event in `EtwEventSource`. For example: `OnEventWritten` would be called right after `FunctionScheduled` or `FunctionAwaited`.

You can read about it [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventsource.loggingeventsource?view=dotnet-plat-ext-3.1) and [here](https://developers.redhat.com/blog/2019/12/23/tracing-net-core-applications/)

_Why does this work?_

**_Currently_**,  our internal logic always calls upon `EtwEventSource` events but, on Linux, they have no effect. Therefore, with the Listener, they would have the effect of invoking the `OnEventWritten` listener which, in turn, would write a log for the linux platform to consume. So we've turned ETW events having no effects, to now having a logging effect..

_Why am I using this approach?_

Well, when we switched to using the `ILogger` abstraction, I thought the vision was to use it as an interface to control **all** logging in the Durable stack. So my understanding was that maybe there was an **EventSource-*implementation* of the ILogger interface that we were going to refactor `EtwEventSource` into. That was what had me confused for so long, because I couldn't find it. Now I realize it probably doesn't exist and that maybe I misunderstood this vision.

I had also read [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventsource.loggingeventsource?view=dotnet-plat-ext-3.1) that `ILogger` has an `EventSource` bridge: `EventListener`. So, after realizing there was no EventSource-implementation of the ILogger interface, I put those two together and _assumed_ that the plan was to use something like this instead :) . Now, I'm not so sure that was the plan either but that's what this thread is for!

(1)

From my answer to (2), hopefully its clear that I was thinking of leaving the **effect-less** ETW events in Linux and to use the Listener as a way to give them a callback that triggers a logging action. If you were to agree with that plan, then the naming would be correct.

Reply:
Thanks @ConnorMcMahon and @davidmrdavid for the explanations. I now understand what `EventListener` does and I think it's actually a better solution than the `ILogger` vision I had originally.

However, regarding (2) I think we still need to remove ETW from the class name. This class doesn't have anything to do with ETW as far as I can tell, so including ETW in the name is misleading. I'm not understanding the point about effect-less ETW events in Linux. I'm not sure how there can be a concept of "effect-less ETW events in Linux if there are no ETW events in Linux. Apologies in advance if I'm misunderstanding something.

Reply:
I agree with the rename to avoid ETW in the name. My initial explanation of  this was poor, as I was still thinking of EventSource in ETW terms. Since .NET Core 2.x, it is more fair to say that EventSource writes system events. On Windows, these are ETW events, on Linux (and macOS?), these are LTTng, but really, the implementation doesn't matter. EventListener is OS-aware enough to listen to these system events regardless, so keeping this abstraction at the layer of "Events" rather than "ETW" is correct.